drafts/bitcartel-payment-disclosure/draft.rst

```diff
@@ -0,0 +1,336 @@
+
+To create a valid Payment Disclosure an implementation must:
+
+#. Check the txid was confirmed in the blockchain
+#. Create a ``PaymentDisclosureKey`` from parameters to RPC call
+#. Use the key to retrieve its value ``PaymentDisclosureInfo`` from storage
+#. Create and populate a ``PaymentDisclosure``
+
+    ::
+    
+        struct PaymentDisclosurePayload {
+            uint8   version;        // 0 = experimental, 1 = first production version, etc.
+            uint256 esk;            // zcash/NoteEncryption.cpp
+            uint256 txid;           // primitives/transaction.h
+            size_t js;              // Index into CTransaction.vjoinsplit
+            uint8_t n;              // Index into JSDescription fields of length ZC_NUM_JS_OUTPUTS
+            std::string message     // parameter to RPC call
+        };
+
+#. Serialize ``PaymentDisclosurePayload`` and sign the raw data using the ``joinSplitPrivKey`` to generate a signature ``payloadSig``.  Sample C++ code to do this:
+
+    ::
+    
+        // Serialize and hash
+        CHashWriter ss(SER_GETHASH, 0);
+        ss << payload << nHashType;
+        uint256 dataToBeSigned = ss.GetHash();
+    
+        // Compute the payload signature
+        unsigned char[64] payloadSig;
+        if (!(crypto_sign_detached(&payloadSig[0], NULL,
+            dataToBeSigned.begin(), 32,
+            joinSplitPrivKey
+            ) == 0))
+        {
+            throw std::runtime_error("crypto_sign_detached failed");
+        }
+    
+        // Sanity check
+        if (!(crypto_sign_verify_detached(&payloadSig[0],
+            dataToBeSigned.begin(), 32,
+            joinSplitPubKey.begin()
+            ) == 0))
+        {
+            throw std::runtime_error("crypto_sign_verify_detached failed");
+        }
+
+
+#. Construct and serialize a PaymentDisclosure object using the data generated so far.
+
+    ::
+    
+        struct PaymentDisclosure {
+            PaymentDisclosurePayload    payload;
+            unsigned char[64]           payloadSig;
+        }
+
+#. Return to the caller a JSON object containing the hex string of the serialized PaymentDisclosure.  If there were errors generating the payment disclosure, return a standard JSON-RPC error with an appropriate error message.
+
+::
+
+    {
+      "hex" : "091234ABCDEF234781248273842738491....."
+    }
+
+
+This raw hex string can be given to the recipient or a third party to use with the new RPC call ``z_validatepaymentdisclosure``:
+
+``z_validatepaymentdisclosure paymentdisclosure``
+
+
+To validate a payment disclosure, perform the following steps:
+
+1. Deserialize the raw hex string into a ``PaymentDisclosure`` object
+2. Retrieve the ``joinSplitPubKey`` for the transaction and verify the payment disclosure signature ``payloadSig``.
+3. Retrieve the note ciphertext from the blockchain for ``txid``, ``js``, ``n``.
+4. Use the ``esk`` to decrypt the ciphertext into plain text
```

Comment:
Does "decrypt" here mean just deriving the symmetric key and decrypting the ciphertext, or the full `DecryptNote` procedure of  [PROTOCOL] §4.10.2, including checking the note commitment?
How does this affect `valid` of the resulting JSON?

Reply:
If the note commitment is note verified, attacks are possible (see discussing starting https://github.com/zcash/zcash/issues/2036#issuecomment-329864311).

Reply:
... and PR #2159 does not verify the note commitment, hence is vulnerable.

Reply:
Updated with a step 5 to verify note commitment.  Good point about "decrypt" and "DecryptNote", will think about this a bit more and perhaps update in a later draft.

Reply:
In general, when specifying the actual algorithms, it would be good to use the same notation and level of formality as in the protocol spec. (For one, this would have caught this bug...)